nibabel/cifti2/cifti2_axes.py

```diff
@@ -0,0 +1,1441 @@
+"""
+Defines :class:`Axis` objects to create, read, and manipulate CIfTI2 files
+
+These axes provide an alternative interface to the information in the CIFTI2 header.
+Each type of CIfTI2 axes describing the rows/columns in a CIfTI2 matrix is given a unique class:
+
+* :class:`BrainModel`: each row/column is a voxel or vertex
+* :class:`Parcels`: each row/column is a group of voxels and/or vertices
+* :class:`Scalar`: each row/column has a unique name (with optional meta-data)
+* :class:`Label`: each row/column has a unique name and label table (with optional meta-data)
+* :class:`Series`: each row/column is a timepoint, which increases monotonically
+
+All of these classes are derived from the :class:`Axis` class.
+
+After loading a CIfTI2 file a tuple of axes describing the rows and columns can be obtained
+from the :meth:`.cifti2.Cifti2Header.get_axis` method on the header object
+(e.g. ``nibabel.load(<filename>).header.get_axis()``). Inversely, a new
+:class:`.cifti2.Cifti2Header` object can be created from existing :class:`Axis` objects
+using the :meth:`.cifti2.Cifti2Header.from_axes` factory method.
+
+CIfTI2 :class:`Axis` objects of the same type can be concatenated using the '+'-operator.
+Numpy indexing also works on axes
+(except for Series objects, which have to remain monotonically increasing or decreasing).
+
+Creating new CIfTI2 axes
+-----------------------
+New :class:`Axis` objects can be constructed by providing a description for what is contained
+in each row/column of the described tensor. For each :class:`Axis` sub-class this descriptor is:
+
+* :class:`BrainModel`: a CIfTI2 structure name and a voxel or vertex index
+* :class:`Parcels`: a name and a sequence of voxel and vertex indices
+* :class:`Scalar`: a name and optionally a dict of meta-data
+* :class:`Label`: a name, dict of label index to name and colour,
+  and optionally a dict of meta-data
+* :class:`Series`: the time-point of each row/column is set by setting the start, stop, size,
+  and unit of the time-series
+
+Several helper functions exist to create new :class:`BrainModel` axes:
+
+* :meth:`BrainModel.from_mask` creates a new BrainModel volume covering the
+  non-zero values of a mask
+* :meth:`BrainModel.from_surface` creates a new BrainModel surface covering the provided
+  indices of a surface
+
+A :class:`Parcels` axis can be created from a sequence of :class:`BrainModel` axes using
+:meth:`Parcels.from_brain_models`.
+
+Examples
+--------
+We can create brain models covering the left cortex and left thalamus using:
+
+>>> from nibabel import cifti2
+>>> bm_cortex = cifti2.BrainModel.from_mask(cortex_mask,
+...                                         brain_structure='cortex_left') # doctest: +SKIP
```

Comment:
Are you able to generate these (or dummy) masks easily? Since we run doctests on docs like this, it would help to ensure that the docs are providing usable suggestions. If not, we can add this to the list of things to look into when we do a big sweep through the docs.

Reply:
I created some small, fake masks. I guess a better example would load some atlas and select a single ROI to include in the BrainModelAxis, but I am not aware of one easily availble